### PR TITLE
LPS-186757 | Expose query parameters in API Explorer for generated API

### PR DESCRIPTION
**Pull request by Hijky**
Background:
Add a testcase to expose query parameters in API Explorer for generated API

Test Plan
CanSeeFilterAndSortParametersInComapnyScopedEndpoint
Given an published API application with company scoped endpoint
When I expand get{endpoint}Page in /o/c/{aPIApplication}
Then I can see filter and sort parameters with string type are present

CanSeeFilterAndSortParametersInSiteScopedEndpoint
Given an published API application with site scoped endpoint
When I expand getscopeScopeKey{Endpoint}Page in /o/c/{aPIApplication}
Then I can see filter and sort parameters with string type are present

CannotSeeFilterAndSortParametersForSingleElementEndpoint
Given an published API application with singleElement endpoint
When I expand get{EndpointPath} in /o/c/{aPIApplication}
Then filter and sort parameters are not present

CanApplyFilterParameter
When execute get{Endpoint}Page with filter={property} eq 'Able' in API Explorer
Then I can see correct response with property equals to Able

CanApplySortParameter
When execute get{Endpoint}Page with sort={property}:desc in API Explorer
Then I can see correct response with items in desc {Test,Tom,Able}

CanApplyFilterParameterAndPreFilter
And Given I create endpoint a filter with name ne 'Test'
When execute get{Endpoint}Page with filter={property} ne 'Able' in API Explorer
Then only Tom returned in response

CanApplySortParameterAndPreSort
And Given I create endpoint a sort with name:asc
When execute get{Endpoint}Page with sort {property}:desc in API Explorer
Then I can see items in desc sort order in response

CanApplyFilterSortAndPreFilterWithSiteScopedEndpoint
Given custom object Student with name text field is created and published
And Given Students Able, Tom and Test are created
And Given an published API application with site scoped endpoint with related schema with Student as the main object   
And Given I create schema a property with erc of name field
And Given I create endpoint a filter with name ne 'Test'
When execute getScopeScopeKey{Endpoint}Page with scopeKey and ?filter={property} ne 'Test'&sort={property}:desc in API Explorer
Then I can get correct response with items in desc {Tom, Able}

Jira ticket:
https://liferay.atlassian.net/browse/LPS-195507

### CODE REVIEW

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/SortAPI.macro [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		return ${curl};
+	}
+
+	macro createAPISort {
+		Variables.assertDefined(parameterList = "${sortValue},${endpointId}");
+
+		var curl = SortAPI._curlAPISort(virtualHost = ${virtualHost});
+		var body = '''
+			"oDataSort": "${sortValue}",
+            "r_apiEndpointToAPISorts_c_apiEndpointId": ${endpointId}
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/SortAPI.macro:32]: sf

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+		if (${testLiferayVirtualInstance} == "true") {
+			PortalInstances.tearDownCP();
+		}
+	}
+
+	@priority = 5
+	test CanApplyFilterParameter {
+		property portal.acceptance = "true";
+
+		task ("Given custom object Student with name text field is created and published") {
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase:32]: Can we put create Student object definition and Student entries(Able, Tom and Test) into Setup()?

Comment (Hijky) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase:32]: but that's not all the case need this,we still add this in setup?

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+		task ("When I expand get{endpoint}Page in /o/c/{aPIApplication}") {
+			APIExplorer.navigateToOpenAPI(customObjectPlural = "my-app");
+
+			APIExplorer.expandAPIMethod(
+				method = "getTestEndpointPage",
+				service = "default");
+		}
+
+		task ("Then I can see filter and sort parameters with string type are present") {
+			AssertElementPresent(
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase:500]: I think it doesn't assert the 'filter' and 'sort' are in 'String' type. Could you please check and update? Thanks.

File: modules/apps/headless/headless-discovery/headless-discovery-test/src/testFunctional/macros/APIExplorer.macro [hunk 4]
```diff
@@ -139,6 +147,15 @@ definition {
 			value = "Execute Query (Ctrl-Enter)");
 	}
 
+	macro expandAPIMethod {
```
Comment (yanan) [inline modules/apps/headless/headless-discovery/headless-discovery-test/src/testFunctional/macros/APIExplorer.macro:150]: Would you please update to expandAPIMethod() in executeAPIMethod() after adding this expandAPIMethod()?

**Review by yanan — CHANGES_REQUESTED** [review]
A few small changes

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase (hunk 2 above)
Comment (Hijky) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase:32]: @magjed4289 what do you think of it?

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase (hunk 2 above)
Comment (Hijky) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/tests/[APIRest]APIGenerator/ExposeQueryParametersInAPIExplorerForGeneratedAPI.testcase:15]: @magjed4289 I have moved the Student object definition and Student entries in Setup()

**Comment by Hijky** [comment]
ci:test:sf

**Comment by Hijky** [comment]
ci:test:headless-functional:nocompile

**Comment by Hijky** [comment]
The failure about ListAPIApplicationEndpoints#CanCopyURLOfAPIEndpoint is a flaky 404 error, the others are [Unable to update user email address](https://liferay.atlassian.net/browse/LPS-198248)

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/ApplicationAPI.macro [hunk 5]
```diff
@@ -57,25 +57,40 @@ definition {
 				var externalReferenceCodeEndpoint = ${endpointName};
 			}
 
-			var relatedEndpoint = '''
+			if (isSet(retrieveType)) {
+				var relatedEndpoint = '''
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/ApplicationAPI.macro:61]: sf

File: modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/ApplicationAPI.macro [hunk 6]
```diff
@@ -57,25 +57,40 @@ definition {
...
+						"path": "${endpointPath}",
+						"pathParameter": "${pathParameter}",
+						"retrieveType": "${retrieveType}",
+						"scope": {"key": "${scopeKey}"}
+					}]
+			''';
+			}
+			else {
+				var relatedEndpoint = '''
+				"apiApplicationToAPIEndpoints": [{
```
Comment (yanan) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testFunctional/macros/ApplicationAPI.macro:75]: sf

**Review by yanan — CHANGES_REQUESTED** [review]

**Review by magjed4289 — APPROVED** [review]

**Comment by yanan** [comment]
ci:forward

**Pull request closed without merging**